gh-18: removes the Batch Credential Endpoint (#364)
openid/OpenID4VCI: openid-4-verifiable-credential-issuance-1_0.md

This specification defines an API for Credential issuance provided by a Credential Issuer. The API is comprised of the following endpoints:

> * A mandatory Credential Endpoint from which Credentials can be issued (see (#credential-endpoint)).

Reviewer: ```suggestion
* A mandatory Credential Endpoint from which Credentials can be issued (see (#credential-endpoint)). From this endpoint, one Credential, or multiple copies of the same Credential Dataset can be issued in one request.
```

Author: I would avoid the term "copy", since I don't think it is used elsewhere in the document.

Author: Changed it to `multiple Credentials with the same Credential Dataset can be issued in one request`

* An optional Deferred Credential Endpoint to allow for the deferred delivery of Credentials (see (#deferred-credential-issuance)).
* An optional mechanism for the Credential Issuer to make a Credential Offer to the Wallet to encourage the Wallet to start the issuance flow (see (#credential-offer-endpoint)).
* An optional mechanism for the Credential Issuer to receive from the Wallet notification(s) of the status of the Credential(s) that have been issued.
* A mechanism for the Credential Issuer to publish metadata about the Credentials it is capable of issuing (see (#credential-issuer-metadata)).

> The Credential Endpoint has the (optional) ability to bind an issued Credential to certain cryptographic key material. Credential requests therefore enable conveying proof of possession for the key material. Multiple key proof types are supported.

Reviewer 2: This is an introduction, I don't think it needs to be normative.

I think the scope of this pull request is removing the batch credential endpoint, and we should not expand the scope of this pull request to introducing normative language where it wasn't present before.

> The Credential Endpoint has the (optional) ability to bind an issued Credential to certain cryptographic key material. Credential requests therefore enable conveying proof of possession for the key material. Multiple key proof types are supported.

Reviewer 3: updated.

### Credential Formats and Credential Format Profiles

This specification is Credential Format agnostic and allows implementers to leverage specific capabilities of Credential Formats of their choice.
To this end, extension points to add Credential Format specific parameters in the Credential Issuer metadata, Credential Offer, Authorization Request, and Credential Request are defined.

Credential Format Profiles for IETF SD-JWT VC [@I-D.ietf-oauth-sd-jwt-vc], ISO mDL [@ISO.18013-5], and W3C VCDM [@VC_DATA] are specified in (#format-profiles).
Other specifications or deployments can define their own Credential Format Profiles using the above-mentioned extension points.

### Multiple Credential Issuance

> This specification allows for the issuance of Verifiable Credentials through the Credential Endpoint.
> A single request message sent to this endpoint requests the issuance one or more Verifiable Credentials.

Reviewer 3: ```suggestion
This specification enables the issuance of Verifiable Credentials through the Credential Endpoint. A single request message to this endpoint MAY request the issuance of one or more Verifiable Credentials.
```

Reviewer: let's not use normative MAYs in a text that is purely meant to be descriptive.

> A single request message sent to this endpoint requests the issuance one or more Verifiable Credentials.

Reviewer: ```suggestion
From this endpoint, one Credential, or multiple copies of the same Credential Dataset can be issued in one request.
```

Credentials can vary in their format, including Credential Format Profile-specific parameters, in their contents known as the Credential Dataset, and in the cryptographic data such as Issuer signatures, hashes, and keys used for Cryptographic Holder Binding.
Multiple Credentials can therefore vary in the following dimensions:

- Credential Dataset
- Credential Format
- Cryptographic Data

> All the credentials issued in the context of a single request MUST have the same Credential Format and Credential Dataset, however they can can have different Cryptographic Data, namely be bound to different cryptographic keys (e.g., for unlinkability between the credentials).

Reviewer 3: ```suggestion
In the context of a single request, all issued Credentials MUST share the same Credential Format and Credential Dataset. However, they MAY contain different Cryptographic Data, such as being bound to different cryptographic keys (e.g., to ensure unlinkability between the credentials).
```

> Issuance of multiple Verifiable Credentials with different Credential Formats or different Credential Datasets requires sending multiple requests to the Credential Endpoint.

Reviewer 3: ```suggestion
To issue multiple Verifiable Credentials with differing Credential Formats or Credential Datasets, multiple requests MUST be sent to the Credential Endpoint.
```

The `proof` element MUST incorporate the Credential Issuer Identifier (audience), and a `c_nonce` value generated by the Authorization Server or the Credential Issuer to allow the Credential Issuer to detect replay. The way that data is incorporated depends on the key proof type. In a JWT, for example, the `c_nonce` value is conveyed in the `nonce` claim, whereas the audience is conveyed in the `aud` claim. In a Linked Data proof, for example, the `c_nonce` is included as the `challenge` element in the key proof object and the Credential Issuer (the intended audience) is included as the `domain` element.

> The initial `c_nonce` value can be returned in a successful Token Response as defined in (#token-response), or in a Credential Error Response as defined in (#issuer-provided-nonce).

Reviewer 3: ```suggestion
The initial `c_nonce` value MAY be returned in a successful Token Response as defined in (#token-response), or in a Credential Error Response as defined in (#issuer-provided-nonce).
```

Reviewer: this is description

# Notification Endpoint {#notification-endpoint}

> This endpoint is used by the Wallet to notify the Credential Issuer of certain events for issued Credentials. These events enable the Credential Issuer to take subsequent actions after issuance. The Credential Issuer needs to return one or more `notification_id` parameters in the Credential Response for the Wallet to be able to use this Endpoint. Support for this endpoint is OPTIONAL. The Issuer cannot assume that a notification will be sent for every issued credential since the use of this Endpoint is not mandatory for the Wallet.

Reviewer 3: ```suggestion
This endpoint is used by the Wallet to notify the Credential Issuer of certain events for issued Credentials. These events enable the Credential Issuer to take subsequent actions after issuance. The Credential Issuer needs to return one or more `notification_id` parameters in the Credential Response for the Wallet to be able to use this endpoint. Support for this endpoint is OPTIONAL. The Issuer cannot assume that a notification will be sent for every issued Credential since the use of this Endpoint is not mandatory for the Wallet.
```

* `credential_endpoint`: REQUIRED. URL of the Credential Issuer's Credential Endpoint, as defined in (#credential-request). This URL MUST use the `https` scheme and MAY contain port, path, and query parameter components.
* `deferred_credential_endpoint`: OPTIONAL. URL of the Credential Issuer's Deferred Credential Endpoint, as defined in (#deferred-credential-issuance). This URL MUST use the `https` scheme and MAY contain port, path, and query parameter components. If omitted, the Credential Issuer does not support the Deferred Credential Endpoint.
* `notification_endpoint`: OPTIONAL. URL of the Credential Issuer's Notification Endpoint, as defined in (#notification-endpoint). This URL MUST use the `https` scheme and MAY contain port, path, and query parameter components. If omitted, the Credential Issuer does not support the Notification Endpoint.

> * `credential_response_encryption`: OPTIONAL. Object containing information about whether the Credential Issuer supports encryption of the Credential Credential Response on top of TLS.

Reviewer 3: ```suggestion
* `credential_response_encryption`: OPTIONAL. Object containing information about whether the Credential Issuer supports encryption of the Credential Credential Response on top of TLS.
```
this "ont op of TLS" sounds weird, we assume that all the requests happens on top of TLS

Reviewer 3: updated, I suggest the removal of "on top of TLS" because in a secure environment all the web services are protected through TLS

Author: Since it was there in the beginning, I prefer to leave it there since it exceeds the PR scope.

  * `alg_values_supported`: REQUIRED. Array containing a list of the JWE [@!RFC7516] encryption algorithms (`alg` values) [@!RFC7518] supported by the Credential Endpoint to encode the Credential Response in a JWT [@!RFC7519].
  * `enc_values_supported`: REQUIRED. Array containing a list of the JWE [@!RFC7516] encryption algorithms (`enc` values) [@!RFC7518] supported by the Credential Endpoint to encode the Credential Response in a JWT [@!RFC7519].
  * `encryption_required`: REQUIRED. Boolean value specifying whether the Credential Issuer requires the additional encryption on top of TLS for the Credential Response. If the value is `true`, the Credential Issuer requires encryption for every Credential Response and therefore the Wallet MUST provide encryption keys in the Credential Request. If the value is `false`, the Wallet MAY chose whether it provides encryption keys or not.

> * `batch_credential_issuance`: OPTIONAL. Object containing information about the Credential Issuer's supports for batch issuance of Credentials on the Credential Endpoint. The presence of this parameter means that the issuer can issue more than one Verifiable Credential for the same Credential Dataset on the same request.

Reviewer 2: I was initially surprised to see credential_endpoint_supported removed, but it makes sense given the presence of this object indicates it. It's probably nice to keep the explicit reference to the `proofs` parameter though:

```suggestion
* `batch_credential_issuance`: OPTIONAL. Object containing information about the Credential Issuer's supports for batch issuance of Credentials on the Credential Endpoint. The presence of this parameter means that the issuer supports the `proofs` parameter in the Credential Request so can issue more than one Verifiable Credential for the same Credential Dataset in a single request/response.
```

Reviewer 3: issuer -> Issuer

After a Verifiable Credential has been issued to the Holder, claim values about the subject of a Credential or a signature on the Credential may need to be updated. There are two possible mechanisms to do so.

> First, the Wallet may receive an updated version of a Credential from a Credential Endpoint using a valid Access Token. This does not involve interaction with the End-User. If the Credential Issuer issued a Refresh Token to the Wallet, the Wallet would obtain a fresh Access Token by making a request to the Token Endpoint, as defined in Section 6 of [@!RFC6749].

Reviewer 3: ```suggestion
Initially, the Wallet MAY receive an updated version of a Credential from a Credential Endpoint using a valid Access Token without involving the End-User. If the Credential Issuer has issued a Refresh Token to the Wallet, the Wallet SHOULD obtain a fresh Access Token by making a request to the Token Endpoint, as specified in Section 6 of [@!RFC6749].
```

Reviewer: this is descriptive text.

To prevent these types of correlation, Credential Issuers and Wallets SHOULD use
methods, including but not limited to the following ones:

> * Issue a batch of Credentials with the same Credential Dataset to enable the usage of a unique Credential per presentation or per Verifier. This only helps with Verifier/Verifier unlinkability.

Reviewer 3: ```suggestion
* Issue a batch of Credentials with the same Credential Dataset to facilitate the use of a unique Credential per presentation or per Verifier. This approach solely aids in achieving Verifier-to-Verifier unlinkability.
```

   * changes proof type descriptions to accommodate for the batch issuance changes
   * fix indentation of examples
   * changes proof type descriptions to accomodate for the batch issuance changes

>    * changed Batch Issuance endpoint to differentiate between issuance Credential Configuration, Credential Dataset and instances of Credentials with different key material

Reviewer: I really don't think anything in the document history should be removed, please rather add one line that you removed batch credential endpoint.

Reviewer 2: As this was a change log entry for the yet-to-be-published draft 14, I believe it is correct to remove it as when we publish -14 it will no longer contain the changes that are being described by this line and hence the only change we need to draw the readers attention to is the removal of the batch endpoint.

(I agree that we should not edit any history entries for published drafts.)

Reviewer: makes sense

   * clarify optionality of scope and `authorization_details` for Authorization Request

> 

Reviewer 2: Delete rather than adding a blank line:

```suggestion
```

   * Define Credential Format as a term
   * Define Credential Dataset as a term
   * Define Credential Configuration as a term

>    * Removes the Batch Credential Endpoint

Reviewer: can you please add this to the top of the history?

Author: Will do.
